cloudinit/stages.py

```diff
@@ -252,12 +252,31 @@ def _get_data_source(self, existing):
             (cfg_list, pkg_list) = self._get_datasources()
             # Deep copy so that user-data handlers can not modify
             # (which will affect user-data handlers down the line...)
-            (ds, dsname) = sources.find_source(self.cfg,
-                                               self.distro,
-                                               self.paths,
-                                               copy.deepcopy(self.ds_deps),
-                                               cfg_list,
-                                               pkg_list, self.reporter)
+            try:
+                (ds, dsname) = sources.find_source(self.cfg,
+                                                   self.distro,
+                                                   self.paths,
+                                                   copy.deepcopy(self.ds_deps),
+                                                   cfg_list,
+                                                   pkg_list, self.reporter)
+            except sources.DataSourceNotFoundException as e:
+                LOG.info("No datasource found - "
+                         "Trying to load previous datasource from cache")
+                ds = self._restore_from_cache()
```

Comment:
I don't understand why ` _restore_from_cache` is being called directly outside `_restore_from_checked_cache`. This entire block will only happen if cloud-init *could not* restore any data from cache (line 250, `if not ds:`). Why you're trying to read it again?

Reply:
@otubo 
This PR I have reopend from #229  as @raharper explained #229 has initial discussion. with this PR I am able to confirm its not resetting to fallback dhcp network when config drive is removed on PowerVM hypervisors setup. 

Even I had to go through the initial change in #229 many times to understand it.  The behavior I am observing when config drive is removed from in my env is  _restore_from_checked_cache returns None saying invalid DS.

"_restore_from_checked_cache is return (None, "cache invalid in datasource: %s" % ds)" (line 262) 

so this would mean probably config drive has gone for PowerVM host as explained  LP: 1893770.  So may be comparing restored id from cache and previous instance id can used to determine if its first boot or subsequent boot here.

As per the documentation in the below link [1] here with "trust" changing to "check" I believe it forces to determine that its first boot or subsequent boot instead of trusting unconditionally. I hope my understanding is right here.

existing = "check"

cloud-init attempts to determine which case it is running in by checking the instance ID in the cache against the instance ID it determines at runtime. If they do not match, then this is an instance’s first boot; otherwise, it’s a subsequent boot.

existing = "trust" 
trust the instance ID that is present in the system unconditionally. This means that cloud-init will never detect a new instance when the cache is present, and it follows that the only way to cause cloud-init to detect a new instance (and therefore its first boot) is to manually remove cloud-init’s cache. 



https://cloudinit.readthedocs.io/en/latest/topics/boot.html

> I don't understand why ` _restore_from_cache` is being called directly outside `_restore_from_checked_cache`. This entire block will only happen if cloud-init _could not_ restore any data from cache (line 250, `if not ds:`). Why you're trying to read it again?

The behavior I am observing when config drive is removed in my env is  _restore_from_checked_cache returns None saying invalid DS as show below in line 262

"_restore_from_checked_cache is return (None, "cache invalid in datasource: %s" % ds)" (line 262) 

so this would mean probably config drive has gone for PowerVM host as explained  LP: 1893770.  So may be comparing restored id from cache and previous instance id can used to determine if its first boot or subsequent boot here.